numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -309,6 +317,85 @@ static int
 
 /**end repeat**/
 
+#ifdef HAVE_STRTOLD_L
+static NPY_INLINE npy_longdouble string_to_long_double(PyObject*op)
+{
+    char *s;
+    char *end;
+    npy_longdouble temp;
+    PyObject* b;
+
+    /* This weird structure is to make the error handling work:
+     * if it's not a string, we want to fall back to other
+     * approaches. In python 2, that happens when we try to
+     * convert the string to a C string. In python 3, we have
+     * to use a two-step conversion process, via bytes; the
+     * not-a-string situation gets triggered on conversion to 
+     * bytes, but we want to make sure bytes objects get 
+     * handled too, because python accepts bytes as input to
+     * float() (why?).
+     */
+#if PY_MAJOR_VERSION >= 3
```

Comment:
Look in `numpy/core/include/numpy/py3_compat.h` for string routines. The ones available are the union of the py3k and py2k functions. My guess you will need to check for unicode and convert it to a byte string, then use `PyString_AS_STRING` to get the c ascii version.

The detection of `strtold_l` would be part of the configuration, which you can find scattered about in `numpy/core/setup_common.py`, `numpy/core/setup.py` and `numpy/core/bscript`. I haven't checked the check function to see if they will work with those particular functions. It might be easier to just copy the functions and make a library, but it looks like even `dtoa.c` has platform dependent bits.